feat: add partial support for WASI

kkebo opened the pull request:
## Description

This change allows building the `cmark-gfm` target for both `wasm32-unknown-wasi` and `wasm32-unknown-wasip1-threads`. The other targets are not tested.

## The current problem

The differences between two target triples are as follows:

- `wasm32-unknown-wasi`
  - C
    - can import `unistd.h`
    - has `__wasi__`
    - does not have `__unix__`
    - **does not have `_REENTRANT`**
    - **does not have `_POSIX_THREADS`**
    - **does not support pthreads API at all**
  - Swift
    - can use `#if os(WASI)`
  - SwiftPM
    - `BuildSettingsCondition`'s `Platform` is `.wasi`
- `wasm32-unknown-wasip1-threads`
  - C
    - can import `unistd.h`
    - has `__wasi__`
    - does not have `__unix__`
    - **has `_REENTRANT` (defined in wasi-libc)**
    - **has `_POSIX_THREADS` (defined in wasi-libc)**
    - **supports a subset of pthreads API**
  - Swift
    - can use `#if os(WASI)`
  - SwiftPM
    - `BuildSettingsCondition`'s `Platform` is `.wasi`

Therefore, `CMARK_THREADING` must not be defined when building for `wasm32-unknown-wasi`, but must be defined when building for `wasm32-unknown-wasip1-threads`.

However, when building with CMake, it is controllable, but when building with SwiftPM, [`CMARK_THREADING` is always defined](https://github.com/swiftlang/swift-cmark/blob/b022b08312decdc46585e0b3440d97f6f22ef703/Package.swift#L18).

## Proposed solution

- Building with CMake and Make
  - Current
    - When running with `-DCMARK_THREADING=OFF` (default), `CMARK_THREADING` will not be defined in the C compiler.
    - When running with `-DCMARK_THREADING=ON`, `CMARK_THREADING` will be defined in the C compiler.
  - New
    - When running with `-DCMARK_THREADING=OFF` (default), `CMARK_THREADING` will be defined in `config.h` and be `0`.
    - When running with `-DCMARK_THREADING=ON`, `CMARK_THREADING` will be defined in `config.h` and be `1`.
- Building with SwiftPM
  - Current: `.define("CMARK_THREADING")` will be added in `Package.swift`.
  - New: `.define("CMARK_THREADING")` will not be added in `Package.swift`. (reverting aef4761a7878a437b981251fc31d0a7b7392de12)
    - Instead, `CMARK_THREADING` will automatically be defined as one of the following values in `src/include/cmark-gfm_config.h`.
      - `0` if building for targets that do not support threading (e.g. `wasm32-unknown-wasi`)
      - `1` if building for the other targets including `wasm32-unknown-wasip1-threads`

I reverted 6f648763b260b38ea6bce7fd0dfa24ea2fa9e4ef because I needed `src/config.h.in` and `src/include/cmark-gfm_config.h`.

## Checks

I've tested this PR (415c3ac674340254a9c703b4860e46aa01159aed) by the following ways on Linux.

```
Linux Raspberry-beetle 6.12.21-v8-16k+ #1868 SMP PREEMPT Mon Mar 31 17:10:57 BST 2025 aarch64 GNU/Linux
```

### Building with SwiftPM

[swift-DEVELOPMENT-SNAPSHOT-2025-03-28-a](https://github.com/apple/swift/releases/tag/swift-DEVELOPMENT-SNAPSHOT-2025-03-28-a):

1. Install swiftly 1.0.0
2. `swiftly install main-snapshot-2025-03-28`
3. `swiftly use main-snapshot-2025-03-28`
4. `swift sdk install https://github.com/swiftwasm/swift/releases/download/swift-wasm-DEVELOPMENT-SNAPSHOT-2025-03-28-a/swift-wasm-DEVELOPMENT-SNAPSHOT-2025-03-28-a-wasm32-unknown-wasi.artifactbundle.zip --checksum c2ab6d3360591d14af05e2bcfcc1ba970f2c3c6862e70b47f36075f846921287` ([instructions](https://github.com/swiftwasm/swift/releases/tag/swift-wasm-DEVELOPMENT-SNAPSHOT-2025-03-28-a))
5. `swift build --swift-sdk wasm32-unknown-wasi --target cmark-gfm`
6. `swift sdk install https://github.com/swiftwasm/swift/releases/download/swift-wasm-DEVELOPMENT-SNAPSHOT-2025-03-28-a/swift-wasm-DEVELOPMENT-SNAPSHOT-2025-03-28-a-wasm32-unknown-wasip1-threads.artifactbundle.zip --checksum 020c7a34a9b5da144f2edbe0b4d4116729ddf6a484674e11357b3fc45b226e86` ([instructions](https://github.com/swiftwasm/swift/releases/tag/swift-wasm-DEVELOPMENT-SNAPSHOT-2025-03-28-a))
7. `swift build --swift-sdk wasm32-unknown-wasip1-threads --target cmark-gfm`
8. `swift build`

```console
$ swift --version
Swift version 6.2-dev (LLVM 6f817b270d9dadc, Swift d36b06747a54689)
Target: aarch64-unknown-linux-gnu
Build config: +assertions
$ swift build --swift-sdk wasm32-unknown-wasi --target cmark-gfm
Building for debugging...
[27/27] Compiling cmark-gfm scanners.c
Build of target: 'cmark-gfm' complete! (2.99s)
$ swift build --swift-sdk wasm32-unknown-wasip1-threads --target cmark-gfm
Building for debugging...
[27/27] Compiling cmark-gfm scanners.c
Build of target: 'cmark-gfm' complete! (2.98s)
$ swift build
Building for debugging...
[42/42] Linking api_test
Build complete! (4.55s)
```

### Building with CMake and Make

<details><summary>-DCMARK_THREADING=ON</summary>

```console
$ cmake -S . -B build -DCMARK_THREADING=ON
26ms 
-- The C compiler identification is GNU 12.2.0
-- The CXX compiler identification is GNU 12.2.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Performing Test HAVE_FLAG_ADDRESS_SANITIZER
-- Performing Test HAVE_FLAG_ADDRESS_SANITIZER - Failed
-- Performing Test HAVE_FLAG_SANITIZE_ADDRESS
-- Performing Test HAVE_FLAG_SANITIZE_ADDRESS - Success
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Success
-- Found Threads: TRUE  
-- Checking _FILE_OFFSET_BITS for large files
-- Checking _FILE_OFFSET_BITS for large files - not needed
-- Found Python3: /usr/bin/python3 (found version "3.11.2") found components: Interpreter 
-- Configuring done
-- Generating done
-- Build files have been written to: /home/kebo/swift-cmark/build
$ cmake --build build
[  2%] Building C object src/CMakeFiles/libcmark-gfm.dir/arena.c.o                                                                                                                       
In file included from /home/kebo/swift-cmark/src/arena.c:6:                                                                                                                              
/home/kebo/swift-cmark/src/include/mutex.h:24:24: warning: ISO C does not allow extra ‘;’ outside of a function [-Wpedantic]                                                             
   24 | CMARK_DEFINE_ONCE(NAME); \                                                                                                                                                       
      |                        ^                                                                                                                                                         
/home/kebo/swift-cmark/src/arena.c:8:1: note: in expansion of macro ‘CMARK_DEFINE_LOCK’                                                                                                  
    8 | CMARK_DEFINE_LOCK(arena)                                                                                                                                                         
      | ^~~~~~~~~~~~~~~~~                                                                                                                                                                
[  4%] Building C object src/CMakeFiles/libcmark-gfm.dir/blocks.c.o                                                                                                                      
[  7%] Building C object src/CMakeFiles/libcmark-gfm.dir/buffer.c.o                                                                                                                      
[  9%] Building C object src/CMakeFiles/libcmark-gfm.dir/cmark.c.o                                                                                                                       
[ 11%] Building C object src/CMakeFiles/libcmark-gfm.dir/cmark_ctype.c.o                                                                                                                 
[ 14%] Building C object src/CMakeFiles/libcmark-gfm.dir/commonmark.c.o                                                                                                                  
[ 16%] Building C object src/CMakeFiles/libcmark-gfm.dir/footnotes.c.o                                                                                                                   
[ 19%] Building C object src/CMakeFiles/libcmark-gfm.dir/houdini_href_e.c.o                                                                                                              
[ 21%] Building C object src/CMakeFiles/libcmark-gfm.dir/houdini_html_e.c.o                                                                                                              
[ 23%] Building C object src/CMakeFiles/libcmark-gfm.dir/houdini_html_u.c.o                                                                                                              
[ 26%] Building C object src/CMakeFiles/libcmark-gfm.dir/html.c.o                                                                                                                        
[ 28%] Building C object src/CMakeFiles/libcmark-gfm.dir/inlines.c.o                                                                                                                     
[ 30%] Building C object src/CMakeFiles/libcmark-gfm.dir/iterator.c.o                                                                                                                    
[ 33%] Building C object src/CMakeFiles/libcmark-gfm.dir/latex.c.o                                                                                                                       
[ 35%] Building C object src/CMakeFiles/libcmark-gfm.dir/linked_list.c.o                                                                                                                 
[ 38%] Building C object src/CMakeFiles/libcmark-gfm.dir/man.c.o                                                                                                                         
[ 40%] Building C object src/CMakeFiles/libcmark-gfm.dir/map.c.o
[ 42%] Building C object src/CMakeFiles/libcmark-gfm.dir/node.c.o
In file included from /home/kebo/swift-cmark/src/node.c:6:
/home/kebo/swift-cmark/src/include/mutex.h:24:24: warning: ISO C does not allow extra ‘;’ outside of a function [-Wpedantic]
   24 | CMARK_DEFINE_ONCE(NAME); \
      |                        ^
/home/kebo/swift-cmark/src/node.c:10:1: note: in expansion of macro ‘CMARK_DEFINE_LOCK’
   10 | CMARK_DEFINE_LOCK(nextflag)
      | ^~~~~~~~~~~~~~~~~
/home/kebo/swift-cmark/src/include/mutex.h:24:24: warning: ISO C does not allow extra ‘;’ outside of a function [-Wpedantic]
   24 | CMARK_DEFINE_ONCE(NAME); \
      |                        ^
/home/kebo/swift-cmark/src/node.c:11:1: note: in expansion of macro ‘CMARK_DEFINE_LOCK’
   11 | CMARK_DEFINE_LOCK(safety)
      | ^~~~~~~~~~~~~~~~~
[ 45%] Building C object src/CMakeFiles/libcmark-gfm.dir/plaintext.c.o
[ 47%] Building C object src/CMakeFiles/libcmark-gfm.dir/plugin.c.o
[ 50%] Building C object src/CMakeFiles/libcmark-gfm.dir/references.c.o
[ 52%] Building C object src/CMakeFiles/libcmark-gfm.dir/registry.c.o
In file included from /home/kebo/swift-cmark/src/registry.c:7:
/home/kebo/swift-cmark/src/include/mutex.h:24:24: warning: ISO C does not allow extra ‘;’ outside of a function [-Wpedantic]
   24 | CMARK_DEFINE_ONCE(NAME); \
      |                        ^
/home/kebo/swift-cmark/src/registry.c:16:1: note: in expansion of macro ‘CMARK_DEFINE_LOCK’
   16 | CMARK_DEFINE_LOCK(extensions);
      | ^~~~~~~~~~~~~~~~~
/home/kebo/swift-cmark/src/registry.c:16:30: warning: ISO C does not allow extra ‘;’ outside of a function [-Wpedantic]
   16 | CMARK_DEFINE_LOCK(extensions);
      |                              ^
[ 54%] Building C object src/CMakeFiles/libcmark-gfm.dir/render.c.o                                                                                                                      
[ 57%] Building C object src/CMakeFiles/libcmark-gfm.dir/scanners.c.o
[ 59%] Building C object src/CMakeFiles/libcmark-gfm.dir/syntax_extension.c.o
[ 61%] Building C object src/CMakeFiles/libcmark-gfm.dir/utf8.c.o                      
[ 64%] Building C object src/CMakeFiles/libcmark-gfm.dir/xml.c.o                                                                                                                         
[ 66%] Linking C static library libcmark-gfm.a                                                                                                                                           
[ 66%] Built target libcmark-gfm                                                                                                                                                         
[ 69%] Building C object extensions/CMakeFiles/libcmark-gfm-extensions.dir/autolink.c.o
[ 71%] Building C object extensions/CMakeFiles/libcmark-gfm-extensions.dir/core-extensions.c.o
/home/kebo/swift-cmark/extensions/core-extensions.c:22:30: warning: ISO C does not allow extra ‘;’ outside of a function [-Wpedantic]
   22 | CMARK_DEFINE_ONCE(registered);                                                      
      |                              ^                                                      
[ 73%] Building C object extensions/CMakeFiles/libcmark-gfm-extensions.dir/ext_scanners.c.o
[ 76%] Building C object extensions/CMakeFiles/libcmark-gfm-extensions.dir/strikethrough.c.o 
[ 78%] Building C object extensions/CMakeFiles/libcmark-gfm-extensions.dir/table.c.o        
[ 80%] Building C object extensions/CMakeFiles/libcmark-gfm-extensions.dir/tagfilter.c.o
[ 83%] Building C object extensions/CMakeFiles/libcmark-gfm-extensions.dir/tasklist.c.o     
[ 85%] Linking C static library libcmark-gfm-extensions.a                                                                                                                                
[ 85%] Built target libcmark-gfm-extensions                                                 
[ 88%] Building C object src/CMakeFiles/cmark-gfm.dir/__/bin/main.c.o
[ 90%] Linking C executable cmark-gfm                                                       
[ 90%] Built target cmark-gfm                                                               
[ 92%] Building CXX object api_test/CMakeFiles/api_test.dir/cplusplus.cpp.o                                                                                                              
[ 95%] Building C object api_test/CMakeFiles/api_test.dir/harness.c.o                                                                                                                    
[ 97%] Building C object api_test/CMakeFiles/api_test.dir/main.c.o
/home/kebo/swift-cmark/api_test/main.c: In function ‘reentrant_parse_inline_ext’:
/home/kebo/swift-cmark/api_test/main.c:1374:32: warning: ISO C forbids conversion of object pointer to function pointer type [-Wpedantic]
 1374 |     reentrant_call_func func = (reentrant_call_func)priv;                     
      |                                ^                                                    
/home/kebo/swift-cmark/api_test/main.c: In function ‘parser_interrupt’:                     
/home/kebo/swift-cmark/api_test/main.c:1398:46: warning: ISO C forbids conversion of function pointer to object pointer type [-Wpedantic]
 1398 |   cmark_syntax_extension_set_private(my_ext, (void *)&run_inner_parser, NULL);                                                                                                   
      |                                              ^
[100%] Linking CXX executable api_test
[100%] Built target api_test
$ cat build/src/config.h
#ifndef CMARK_CONFIG_H
#define CMARK_CONFIG_H

#ifdef __cplusplus
extern "C" {
#endif

#define CMARK_THREADING 1

#ifdef __cplusplus
}
#endif

#endif
```

</details>

<details><summary>-DCMARK_THREADING=OFF</summary>

```console
$ cmake -S . -B build -DCMARK_THREADING=OFF
-- The C compiler identification is GNU 12.2.0
-- The CXX compiler identification is GNU 12.2.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Performing Test HAVE_FLAG_ADDRESS_SANITIZER
-- Performing Test HAVE_FLAG_ADDRESS_SANITIZER - Failed
-- Performing Test HAVE_FLAG_SANITIZE_ADDRESS
-- Performing Test HAVE_FLAG_SANITIZE_ADDRESS - Success
-- Checking _FILE_OFFSET_BITS for large files
-- Checking _FILE_OFFSET_BITS for large files - not needed
-- Found Python3: /usr/bin/python3 (found version "3.11.2") found components: Interpreter 
-- Configuring done
-- Generating done
-- Build files have been written to: /home/kebo/swift-cmark/build
$ cmake --build build
[  2%] Building C object src/CMakeFiles/libcmark-gfm.dir/arena.c.o
[  4%] Building C object src/CMakeFiles/libcmark-gfm.dir/blocks.c.o
[  7%] Building C object src/CMakeFiles/libcmark-gfm.dir/buffer.c.o
[  9%] Building C object src/CMakeFiles/libcmark-gfm.dir/cmark.c.o
[ 11%] Building C object src/CMakeFiles/libcmark-gfm.dir/cmark_ctype.c.o
[ 14%] Building C object src/CMakeFiles/libcmark-gfm.dir/commonmark.c.o
[ 16%] Building C object src/CMakeFiles/libcmark-gfm.dir/footnotes.c.o
[ 19%] Building C object src/CMakeFiles/libcmark-gfm.dir/houdini_href_e.c.o
[ 21%] Building C object src/CMakeFiles/libcmark-gfm.dir/houdini_html_e.c.o
[ 23%] Building C object src/CMakeFiles/libcmark-gfm.dir/houdini_html_u.c.o
[ 26%] Building C object src/CMakeFiles/libcmark-gfm.dir/html.c.o
[ 28%] Building C object src/CMakeFiles/libcmark-gfm.dir/inlines.c.o
[ 30%] Building C object src/CMakeFiles/libcmark-gfm.dir/iterator.c.o
[ 33%] Building C object src/CMakeFiles/libcmark-gfm.dir/latex.c.o
[ 35%] Building C object src/CMakeFiles/libcmark-gfm.dir/linked_list.c.o
[ 38%] Building C object src/CMakeFiles/libcmark-gfm.dir/man.c.o
[ 40%] Building C object src/CMakeFiles/libcmark-gfm.dir/map.c.o
[ 42%] Building C object src/CMakeFiles/libcmark-gfm.dir/node.c.o
[ 45%] Building C object src/CMakeFiles/libcmark-gfm.dir/plaintext.c.o
[ 47%] Building C object src/CMakeFiles/libcmark-gfm.dir/plugin.c.o
[ 50%] Building C object src/CMakeFiles/libcmark-gfm.dir/references.c.o
[ 52%] Building C object src/CMakeFiles/libcmark-gfm.dir/registry.c.o
/home/kebo/swift-cmark/src/registry.c:16:30: warning: ISO C does not allow extra â;â outside of a function [-Wpedantic]
   16 | CMARK_DEFINE_LOCK(extensions);
      |                              ^
[ 54%] Building C object src/CMakeFiles/libcmark-gfm.dir/render.c.o
[ 57%] Building C object src/CMakeFiles/libcmark-gfm.dir/scanners.c.o
[ 59%] Building C object src/CMakeFiles/libcmark-gfm.dir/syntax_extension.c.o
[ 61%] Building C object src/CMakeFiles/libcmark-gfm.dir/utf8.c.o
[ 64%] Building C object src/CMakeFiles/libcmark-gfm.dir/xml.c.o
[ 66%] Linking C static library libcmark-gfm.a
[ 66%] Built target libcmark-gfm
[ 69%] Building C object extensions/CMakeFiles/libcmark-gfm-extensions.dir/autolink.c.o
[ 71%] Building C object extensions/CMakeFiles/libcmark-gfm-extensions.dir/core-extensions.c.o
/home/kebo/swift-cmark/extensions/core-extensions.c:22:30: warning: ISO C does not allow extra â;â outside of a function [-Wpedantic]
   22 | CMARK_DEFINE_ONCE(registered);
      |                              ^
[ 73%] Building C object extensions/CMakeFiles/libcmark-gfm-extensions.dir/ext_scanners.c.o
[ 76%] Building C object extensions/CMakeFiles/libcmark-gfm-extensions.dir/strikethrough.c.o
[ 78%] Building C object extensions/CMakeFiles/libcmark-gfm-extensions.dir/table.c.o
[ 80%] Building C object extensions/CMakeFiles/libcmark-gfm-extensions.dir/tagfilter.c.o
[ 83%] Building C object extensions/CMakeFiles/libcmark-gfm-extensions.dir/tasklist.c.o
[ 85%] Linking C static library libcmark-gfm-extensions.a
[ 85%] Built target libcmark-gfm-extensions
[ 88%] Building C object src/CMakeFiles/cmark-gfm.dir/__/bin/main.c.o
[ 90%] Linking C executable cmark-gfm
[ 90%] Built target cmark-gfm
[ 92%] Building CXX object api_test/CMakeFiles/api_test.dir/cplusplus.cpp.o
[ 95%] Building C object api_test/CMakeFiles/api_test.dir/harness.c.o
[ 97%] Building C object api_test/CMakeFiles/api_test.dir/main.c.o
/home/kebo/swift-cmark/api_test/main.c: In function âreentrant_parse_inline_extâ:
/home/kebo/swift-cmark/api_test/main.c:1374:32: warning: ISO C forbids conversion of object pointer to function pointer type [-Wpedantic]
 1374 |     reentrant_call_func func = (reentrant_call_func)priv;
      |                                ^
/home/kebo/swift-cmark/api_test/main.c: In function âparser_interruptâ:
/home/kebo/swift-cmark/api_test/main.c:1398:46: warning: ISO C forbids conversion of function pointer to object pointer type [-Wpedantic]
 1398 |   cmark_syntax_extension_set_private(my_ext, (void *)&run_inner_parser, NULL);
      |                                              ^
[100%] Linking CXX executable api_test
[100%] Built target api_test
$ cat build/src/config.h
#ifndef CMARK_CONFIG_H
#define CMARK_CONFIG_H

#ifdef __cplusplus
extern "C" {
#endif

#define CMARK_THREADING 0

#ifdef __cplusplus
}
#endif

#endif
```

</details>

### Building as an external dependency with SwiftPM

This CI workflow builds swift-format's executable and test executable for `wasm32-unknown-wasi`. swift-format depends on swift-cmark, and the following one is using this PR's branch. The test executable runs on a WebAssembly runtime (wasmtime). As you can see, all tests cases have passed.

~~https://github.com/kkebo/swift-format/actions/runs/11669491583~~
https://github.com/kkebo/swift-format/actions/runs/14224215545

kkebo committed:
feat: add partial support for WASI

This change allows building the `cmark-gfm` target for both `wasm32-unknown-wasi` and `wasm32-unknown-wasip1-threads`. The other targets are not tested.

The differences between two target triples:

- `wasm32-unknown-wasi`
  - C
    - can import `unistd.h`
    - has `__wasi__`
    - does not have `__unix__`
    - **does not have `_REENTRANT`**
    - **does not have `_POSIX_THREADS`**
    - **does not support pthreads API at all**
  - Swift
    - can use `#if os(WASI)`
  - SwiftPM
    - `BuildSettingsCondition`'s `Platform` is `.wasi`
- `wasm32-unknown-wasip1-threads`
  - C
    - can import `unistd.h`
    - has `__wasi__`
    - does not have `__unix__`
    - **has `_REENTRANT` (defined in wasi-libc)**
    - **has `_POSIX_THREADS` (defined in wasi-libc)**
    - **supports a subset of pthreads API**
  - Swift
    - can use `#if os(WASI)`
  - SwiftPM
    - `BuildSettingsCondition`'s `Platform` is `.wasi`

QuietMisdreavus commented:
I feel like it would be nice to still have the ability to definitively disable the threading support, since upstream `cmark-gfm` doesn't have it. What do you think of adding a kind of `CMARK_NO_THREADING` option that can be set from CMake? The top-level `CMakeLists.txt` would be updated like this:

```patch
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 23eca3c5..4f741228 100755
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -24,7 +24,8 @@ set(CMAKE_INCLUDE_CURRENT_DIR YES)
 
 option(CMARK_FUZZ_QUADRATIC "Build quadratic fuzzing harness" OFF)
 option(CMARK_LIB_FUZZER "Build libFuzzer fuzzing harness" OFF)
-option(CMARK_THREADING "Add locks around static accesses" OFF)
+option(CMARK_THREADING "Add locks around static accesses; will autodetect based on platform if not set" OFF)
+option(CMARK_NO_THREADING "Disable thread-support auto-detection" OFF)
 
 if("${CMAKE_SOURCE_DIR}" STREQUAL "${CMAKE_BINARY_DIR}")
     message(FATAL_ERROR "Do not build in-source.\nPlease remove CMakeCache.txt and the CMakeFiles/ directory.\nThen: mkdir build ; cd build ; cmake .. ; make")
@@ -44,6 +45,9 @@ if(CMARK_THREADING)
   include(FindThreads)
   add_compile_definitions(CMARK_THREADING)
 endif()
+if(CMARK_NO_THREADING)
+  add_compile_definitions(CMARK_NO_THREADING)
+endif()
 include(GNUInstallDirs)
 
 if(NOT MSVC OR CMAKE_HOST_SYSTEM_NAME STREQUAL Windows)
```

And then you'd guard the autodetect you added with a `#if !defined(CMARK_THREADING) && !defined(CMARK_NO_THREADING)` so that it's still enabled by default (including when built as a Swift package), but can be explicitly disabled as well if desired.

kkebo commented:
I would like to hear @kateinoigakukun's opinion too, if possible.

kateinoigakukun commented:
@kkebo I'm personally not a big fan of undefining a definition as I feel "including a header results in disappearing a definition" is a bit confusing.

How about interpreting `CMARK_THREADING` definition as a value with `#if` instead of just testing its existence.
In CMake build, we can just propagate `CMARK_THREADING` CMake variable value, and in SwiftPM build, we will get a way to explicitly turn off the support by `swift build -Xcc -DCMARK_THREADING=0`.

kkebo commented:
Thanks for the advice. It seems to be possible to express the three cases (on, off, auto) in that way.

- CMake
  - Current
    - When running with `-DCMARK_THREADING=OFF` (default), `CMARK_THREADING` will not be defined in C.
    - When running with `-DCMARK_THREADING=ON`, `CMARK_THREADING` will be defined in C.
  - New
    - When running with `-DCMARK_THREADING=OFF` (default), `CMARK_THREADING` will be `0` in C.
    - When running with `-DCMARK_THREADING=ON`, `CMARK_THREADING` will be `1` in C.
- SwiftPM
  - Current: `CMARK_THREADING` will be defined in C.
  - New: `CMARK_THREADING` will not be defined in C.

In C,

- Current
  - When `CMARK_THREADING` is not defined, it will fall back to the branch for no threading support.
  - When `CMARK_THREADING` is is defined, it will execute the branch for threading support.
- New
  - When `CMARK_THREADING` is `0`, it will fall back to the branch for no threading support.
  - When `CMARK_THREADING` is `1`, it will execute the branch for threading support.
  - When `CMARK_THREADING` is not defined,
    - If building for targets that do not support threading (e.g. `wasm32-unknown-wasi`), `CMARK_THREADING` will be defined and be `0` when including `mutex.h`.
    - If building for the other targets including `wasm32-unknown-wasip1-threads`, `CMARK_THREADING` will be defined and be `1` when including `mutex.h`.

I felt that this method seemed to be the best at the moment.

<details><summary>new diff from gfm branch</summary>

```diff
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 23eca3c5..02596d5d 100755
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -42,7 +42,9 @@ include(FindAsan)
 if(CMARK_THREADING)
   set(THREADS_PREFER_PTHREAD_FLAG YES)
   include(FindThreads)
-  add_compile_definitions(CMARK_THREADING)
+  add_compile_definitions(CMARK_THREADING=1)
+else()
+  add_compile_definitions(CMARK_THREADING=0)
 endif()
 include(GNUInstallDirs)
 
diff --git a/Package.swift b/Package.swift
index f1acc5f3..618c280a 100644
--- a/Package.swift
+++ b/Package.swift
@@ -11,12 +11,9 @@ import PackageDescription
     // link time.
     let cSettings: [CSetting] = [
         .define("CMARK_GFM_STATIC_DEFINE", .when(platforms: [.windows])),
-        .define("CMARK_THREADING"),
     ]
 #else
-    let cSettings: [CSetting] = [
-        .define("CMARK_THREADING"),
-    ]
+    let cSettings: [CSetting] = []
 #endif
 
 let package = Package(
diff --git a/api_test/harness.c b/api_test/harness.c
index f561aeac..a00ae9ef 100644
--- a/api_test/harness.c
+++ b/api_test/harness.c
@@ -54,7 +54,7 @@ void INT_EQ(test_batch_runner *runner, int got, int expected, const char *msg,
   }
 }
 
-#ifndef _WIN32
+#if !defined(_WIN32) && !defined(__wasi__)
 #include <unistd.h>
 
 static char *write_tmp(char const *header, char const *data) {
@@ -79,7 +79,7 @@ void STR_EQ(test_batch_runner *runner, const char *got, const char *expected,
   va_end(ap);
 
   if (!cond) {
-#ifndef _WIN32
+#if !defined(_WIN32) && !defined(__wasi__)
     char *got_fn = write_tmp("actual\n", got);
     char *expected_fn = write_tmp("expected\n", expected);
     char buf[1024];
diff --git a/src/include/mutex.h b/src/include/mutex.h
index 59427f34..25630622 100644
--- a/src/include/mutex.h
+++ b/src/include/mutex.h
@@ -3,9 +3,17 @@
 
 #include <stdbool.h>
 
-#ifdef CMARK_THREADING
+#ifndef CMARK_THREADING
+#if defined(__wasi__) && !defined(_REENTRANT)
+#define CMARK_THREADING 0
+#else
+#define CMARK_THREADING 1
+#endif
+#endif
+
+#if CMARK_THREADING
 
-#if defined(__unix__) || (defined(__APPLE__) && defined(__MACH__))
+#if defined(__unix__) || (defined(__APPLE__) && defined(__MACH__)) || defined(__wasi__)
 #include <unistd.h>
 #endif
```

</details>

In addition to successful builds for `wasm32-unknown-wasi` and `wasm32-unknown-wasip1-threads`, I've tested CMake too.

<details><summary>CMake logs</summary>

```console
$ cmake -S . -B build -DCMARK_THREADING=OFF
-- Configuring done
-- Generating done
-- Build files have been written to: /home/kebo/swift-cmark/build
$ grep CMARK_THREADING build
build/extensions/CMakeFiles/libcmark-gfm-extensions.dir/flags.make
5:C_DEFINES = -DCMARK_GFM_STATIC_DEFINE -DCMARK_THREADING=0

build/CMakeCache.txt
285:CMARK_THREADING:BOOL=OFF

build/api_test/CMakeFiles/api_test.dir/flags.make
6:C_DEFINES = -DCMARK_GFM_STATIC_DEFINE -DCMARK_THREADING=0
12:CXX_DEFINES = -DCMARK_GFM_STATIC_DEFINE -DCMARK_THREADING=0

build/src/CMakeFiles/libcmark-gfm.dir/flags.make
5:C_DEFINES = -DCMARK_GFM_STATIC_DEFINE -DCMARK_THREADING=0

build/src/CMakeFiles/cmark-gfm.dir/flags.make
5:C_DEFINES = -DCMARK_GFM_STATIC_DEFINE -DCMARK_THREADING=0
$ cmake -S . -B build -DCMARK_THREADING=ON
-- Configuring done
-- Generating done
-- Build files have been written to: /home/kebo/swift-cmark/build
$ grep CMARK_THREADING build
build/extensions/CMakeFiles/libcmark-gfm-extensions.dir/flags.make
5:C_DEFINES = -DCMARK_GFM_STATIC_DEFINE -DCMARK_THREADING=1

build/CMakeCache.txt
285:CMARK_THREADING:BOOL=ON

build/api_test/CMakeFiles/api_test.dir/flags.make
6:C_DEFINES = -DCMARK_GFM_STATIC_DEFINE -DCMARK_THREADING=1
12:CXX_DEFINES = -DCMARK_GFM_STATIC_DEFINE -DCMARK_THREADING=1

build/src/CMakeFiles/libcmark-gfm.dir/flags.make
5:C_DEFINES = -DCMARK_GFM_STATIC_DEFINE -DCMARK_THREADING=1

build/src/CMakeFiles/cmark-gfm.dir/flags.make
5:C_DEFINES = -DCMARK_GFM_STATIC_DEFINE -DCMARK_THREADING=1
```

</details>

kkebo commented:
@QuietMisdreavus If this change addresses your concerns, I would change the PR to this.

compnerd commented:
This would mean that the library built with the C compiler could have threading enabled, but then when imported into swift would not have threading enabled. This feels like it would cause issues. Perhaps we need to consider having a `config.h` generated instead. That could be checked in for SPM usage.

kkebo committed:
Revert "remove empty `config.h` and `cmark-gfm_config.h` headers"

This reverts commit 6f648763b260b38ea6bce7fd0dfa24ea2fa9e4ef.

kkebo committed:
Revert "build: hoist `CMARK_THREADING` to a build-only config"

This reverts commit aef4761a7878a437b981251fc31d0a7b7392de12.

kkebo committed:
build: change `CMARK_THREADING` to 0 or 1

kkebo commented:
@compnerd Thank you for your comment. It would certainly be better to use `config.h` to make the `CMARK_THREADING` value persistent. I've pushed commits and rewritten the description.

I referred to previous commits and https://github.com/github/cmark-gfm.

In addition, as it seemed more appropriate to place the code for when `CMARK_THREADING` is not defined in `cmark-gfm_config.h` rather than in `mutex.h`, I changed it.

compnerd reviewed (commented):
This seems reasonable to me. Please get @QuietMisdreavus to sign off on this.

kkebo commented:
@QuietMisdreavus Are there any concerns remaining?

kkebo commented:
@QuietMisdreavus I've added the ability to definitively disable the threading support to `CMakeLists.txt` as you commented first. Could you please review this again? Thanks.

QuietMisdreavus commented:
GitHub Actions are not enabled on swift-cmark; those configuration files were inherited from GitHub-Flavored Markdown and won't affect anything here.

Thanks for your patience, this kept falling off my radar and/or getting sidetracked by other priorities. Let me see if i can take another look.

QuietMisdreavus commented on src/include/mutex.h:
```diff
@@ -1,11 +1,13 @@
+#if CMARK_THREADING
 
-#if defined(__unix__) || (defined(__APPLE__) && defined(__MACH__))
+#if defined(__unix__) || (defined(__APPLE__) && defined(__MACH__)) || defined(__wasi__)
 #include <unistd.h>
 #endif
```
At this point i'm starting to wonder if we can just get away with just checking for the existence of `unistd.h` instead of probing for different compilation environments:

```suggestion
#if defined(__has_include)
#if __has_include(<unistd.h>)
#include <unistd.h>
#endif
#endif
```

It looks like both Clang and GCC have the `__has_include` macro, so this should probably cover our bases for any other new platform over time.

kkebo replied:
@QuietMisdreavus I really appreciate your review.

As you said, using the `__has_include` macro makes the code more robust against future changes, but I hesitate to use it.

Because the `__has_include` macro is standard in C++17 and C23, but not in C99. This repository appears to be compiled as C99, at least in CMake.

https://github.com/swiftlang/swift-cmark/blob/b97d09472e847a416629f026eceae0e2afcfad65/CMakeLists.txt#L16-L18

QuietMisdreavus replied:
Oh heck, i didn't realize it was tied to a C standard. It's likely the C99 requirement is inherited from GitHub-Flavored Markdown (or even CommonMark) so i would hesitate to change it.

In that case, we can merge this as-is and live with the ever-growing list of Unix-like platforms here. 😅

QuietMisdreavus reviewed (approved):
Took another look. Just to restate the summary of the behavior this introduces:

- When building swift-cmark with CMake (i.e. in the compiler, sometimes at desk when i want to run the full test suite) the default is to disable threading unless otherwise forced with `-DCMARK_THREADING=ON`. It's up to integrators and individual developers to turn it on based on the capabilities of their system. In addition, threading is either "always on" or "always off", since there's no way to "unset" a boolean option and unlock the "automatic" mode.
- When building swift-cmark with SwiftPM (i.e. as a dependency of Swift-Markdown and Swift-DocC), the default is to enable threading *except explicitly on WASI without threads*. Given that the previous default in SwiftPM was "on all the time", this isn't necessarily a bad thing, but it kinda makes me wonder about other future platforms where we would want to turn off threading.

I think i'm fine with this outcome. It allows swift-cmark to automatically turn off threading support when building in WASI without the threading runtime, but still allows CMake builds to definitively turn it on or off. (I do wish there was a better way to do an automatic selection, but i'd rather not try to open up another can of worms and delay this PR for another year. 😅)

I have one source comment, but otherwise i think this is fine to land. I'll wait for your reply (and/or compnerd's reply if he has a better picture) before merging, just in case we need to keep that code as-is.

Thanks so much for putting this up, and for your patience with my reviews! I'm really glad to be able to expand this code to more places.

QuietMisdreavus commented:
Thanks again for the PR!
